lib/logstash/bundler.rb

```diff
@@ -0,0 +1,125 @@
+      end
+    end
+
+    # execute bundle install and capture any $stdout output. any raised exception in the process will be trapped
+    # and returned. logs errors to $stdout.
+    # @param options [Hash] invoke options with default values, :max_tries => 10, :clean => false, :install => false, :update => false
+    # @param   options[:update] must be either false or a String or an Array of String
+    # @return [String, Exception] the installation captured output and any raised exception or nil if none
+    def self.invoke_bundler!(options = {})
+      options = {:max_tries => 10, :clean => false, :install => false, :update => false, :without => [:development]}.merge(options)
+      options[:without] = Array(options[:without])
+      options[:update] = Array(options[:update]) if options[:update]
+
+      ENV["GEM_PATH"] = LogStash::Environment.logstash_gem_home
+      ENV["BUNDLE_PATH"] = LogStash::Environment.logstash_gem_home
+      ENV["BUNDLE_GEMFILE"] = LogStash::Environment::GEMFILE_PATH
+      ENV["BUNDLE_WITHOUT"] = options[:without].empty? ? "" : options[:without].join(":")
+
+      try = 0
+
+      # capture_stdout also traps any raised exception and pass them back as the function return [output, exception]
+      capture_stdout do
+        loop do
+          begin
+            ::Bundler.reset!
+            ::Bundler::CLI.start(bundler_arguments(options))
+            break
+          rescue ::Bundler::VersionConflict => e
+            $stderr.puts("Plugin version conflict, aborting")
+            raise(e)
+          rescue ::Bundler::GemNotFound => e
+            $stderr.puts("Plugin not found, aborting")
+            raise(e)
+          rescue => e
+            if try >= options[:max_tries]
+              $stderr.puts("Too many retries, aborting, caused by #{e.class}")
+              $stderr.puts(e.message) if ENV["DEBUG"]
+              raise(e)
+            end
+
+            try += 1
+            $stderr.puts("Error #{e.class}, retrying #{try}/#{options[:max_tries]}")
+            $stderr.puts(e.message) if ENV["DEBUG"]
+            sleep(0.5)
+          end
+        end
+      end
+    end
+
+    # build Bundler::CLI.start arguments array from the given options hash
+    # @param option [Hash] the invoke_bundler! options hash
+    # @return [Array<String>] Bundler::CLI.start string arguments array
+    def self.bundler_arguments(options = {})
+      arguments = []
+
+      if options[:install]
+        arguments << "install"
+        arguments << "--gemfile=#{LogStash::Environment::GEMFILE_PATH}"
+        arguments << ["--path", LogStash::Environment::BUNDLE_DIR]
+        # note that generating "--without=" when options[:without] is empty is intended
+        arguments << "--without=#{options[:without].join(' ')}"
+      end
+
+      if options[:update]
+        arguments << "update"
+        arguments << options[:update]
+      end
+
+      arguments << "--clean" if options[:clean]
+
+      arguments.flatten
```

Comment:
nope. I AGREE THIS CODE NEEDS REFACTORING ;)  but `:without` and `:update` are arrays.